src/bcf/report/static_reader.rs

```diff
@@ -0,0 +1,170 @@
+    const FLAG_9: &'static str = "secondary alignment";
+    const FLAG_10: &'static str = "not passing filters, such as platform/vendor quality controls";
+    const FLAG_11: &'static str = "PCR or optical duplicate";
+    const FLAG_12: &'static str = "vega lite lines";
+
+    let mut flags_map = BTreeMap::new();
+    flags_map.insert(0x1, FLAG_1);
+    flags_map.insert(0x2, FLAG_2);
+    flags_map.insert(0x4, FLAG_3);
+    flags_map.insert(0x8, FLAG_4);
+    flags_map.insert(0x10, FLAG_5);
+    flags_map.insert(0x20, FLAG_6);
+    flags_map.insert(0x40, FLAG_7);
+    flags_map.insert(0x80, FLAG_8);
+    flags_map.insert(0x100, FLAG_9);
+    flags_map.insert(0x200, FLAG_10);
+    flags_map.insert(0x400, FLAG_11);
+    flags_map.insert(0x800, FLAG_12);
+
+    for (flag, text) in flags_map {
+        for f in flag_vec.clone() {
+            if (flag & f) == flag {
+                string_map.insert(flag, text);
+            }
+        }
+    }
+
+    string_map
+}
+
+fn calc_rows(
+    reads: Vec<AlignmentNucleobase>,
+    matches: Vec<AlignmentMatch>,
+) -> (Vec<StaticAlignmentNucleobase>, Vec<StaticAlignmentMatch>) {
+    let mut row_ends = vec![0; 30];
+
+    let mut read_names: BTreeMap<String, u8> = BTreeMap::new();
+
+    let mut reads_wr: Vec<StaticAlignmentNucleobase> = Vec::new();
+    let mut matches_wr: Vec<StaticAlignmentMatch> = Vec::new();
+
+    for r in matches {
+        let mut row: u8 = 0;
+
+        if read_names.contains_key(&r.name) {
+            row = *read_names.get(&r.name).unwrap();
+        } else {
+            for i in 1..30 {
+                if r.read_start > row_ends[i] {
+                    row = i as u8;
+                    row_ends[i] = r.read_end;
+                    read_names.insert(r.name.clone(), i as u8);
+                    break;
+                }
+            }
+        }
+
+        let f = decode_static_flags(r.flags);
+
+        let base = StaticAlignmentMatch {
+            marker_type: r.marker_type,
+            start_position: r.start_position,
+            end_position: r.end_position,
+            flags: f,
+            name: r.name,
+            row: row,
+        };
+
+        matches_wr.push(base);
+    }
+
+    for r in reads {
+        let mut row: u8 = 0;
+
+        if read_names.contains_key(&r.name) {
+            row = *read_names.get(&r.name).unwrap();
+        } else {
+            for i in 1..30 {
+                if r.read_start > row_ends[i] {
+                    row = i as u8;
+                    row_ends[i] = r.read_end;
+                    read_names.insert(r.name.clone(), i as u8);
+                    break;
+                }
+            }
+        }
+
+        let f = decode_static_flags(r.flags);
+
+        let base = StaticAlignmentNucleobase {
+            marker_type: r.marker_type,
+            bases: r.bases,
+            start_position: r.start_position,
+            end_position: r.end_position,
+            flags: f,
```

Comment:
Does that mean that the flags string ends up in each of the alignment intervals? This sounds like wasting bits. Why not just store the flag ids in JSON, and define a mapping in the vega json?

Reply:
How would such a mapping be implemented in the vega json? Just storing the ids in the JSON would indeed reduce the data that is written into the html file. In the client-server version this is somehow already done. The json array of the flag ids is replaced with the actual flag string with javascript.